File: backend/src/graduate_entrance/schemas/essay.py
from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, Field


class EssayCategory(StrEnum):
    phrase = "phrase"
    sentence = "sentence"
    paragraph = "paragraph"
    template = "template"
    quote = "quote"


class ReciteResult(StrEnum):
    remembered = "remembered"
    forgot = "forgot"


class EssayMaterialCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=240)
    category: EssayCategory = EssayCategory.sentence
    topic: str = Field(default="", max_length=120)
    content_md: str = Field(min_length=1)
    translation_md: str = ""
    source: str = Field(default="", max_length=240)


class EssayMaterialUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=240)
    category: EssayCategory | None = None
    topic: str | None = Field(default=None, max_length=120)
    content_md: str | None = Field(default=None, min_length=1)
    translation_md: str | None = None
    source: str | None = Field(default=None, max_length=240)


class EssayMaterialRead(BaseModel):
    id: UUID
    title: str
    category: EssayCategory
    topic: str
    content_md: str
    translation_md: str
    source: str
    due_date: date | None
    interval_days: int
    recite_count: int
    created_at: datetime
    updated_at: datetime


class EssayMaterialListResponse(BaseModel):
    total: int
    materials: list[EssayMaterialRead]


class ReciteRequest(BaseModel):
    result: ReciteResult


class ReciteResponse(BaseModel):
    material: EssayMaterialRead
    next_due: date

File: web/src/services/essay.ts
import { apiFetch } from './api'

export type EssayCategory = 'phrase' | 'sentence' | 'paragraph' | 'template' | 'quote'
export type ReciteResult = 'remembered' | 'forgot'

export interface EssayMaterial {
  id: string
  title: string
  category: EssayCategory
  topic: string
  content_md: string
  translation_md: string
  source: string
  due_date: string | null
  interval_days: number
  recite_count: number
  created_at: string
  updated_at: string
}

export interface EssayMaterialList {
  total: number
  materials: EssayMaterial[]
}

export interface EssayMaterialInput {
  title: string
  category: EssayCategory
  topic: string
  content_md: string
  translation_md: string
  source: string
}

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(`/api${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  })
  if (!response.ok) {
    throw new Error(`Essay request failed with status ${response.status}`)
  }
  if (response.status === 204) {
    return undefined as T
  }
  return (await response.json()) as T
}

export function fetchMaterials(options: {
  category?: EssayCategory
  q?: string
  dueOnly?: boolean
} = {}): Promise<EssayMaterialList> {
  const params = new URLSearchParams()
  if (options.category) {
    params.set('category', options.category)
  }
  if (options.q) {
    params.set('q', options.q)
  }
  if (options.dueOnly) {
    params.set('due_only', 'true')
  }
  const query = params.toString()
  return requestJson<EssayMaterialList>(`/essay/materials${query ? `?${query}` : ''}`)
}

export function createMaterial(input: EssayMaterialInput): Promise<EssayMaterial> {
  return requestJson<EssayMaterial>('/essay/materials', {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

export function updateMaterial(
  id: string,
  input: Partial<EssayMaterialInput>,
): Promise<EssayMaterial> {
  return requestJson<EssayMaterial>(`/essay/materials/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(input),
  })
}

export function deleteMaterial(id: string): Promise<void> {
  return requestJson<void>(`/essay/materials/${id}`, { method: 'DELETE' })
}

export function reciteMaterial(
  id: string,
  result: ReciteResult,
): Promise<{ material: EssayMaterial; next_due: string }> {
  return requestJson<{ material: EssayMaterial; next_due: string }>(
    `/essay/materials/${id}/recite`,
    {
      method: 'POST',
      body: JSON.stringify({ result }),
    },
  )
}
